src/manifolds/StiefelCanonicalMetric.jl

```diff
@@ -186,3 +174,52 @@ function inverse_retract!(
     @views mul!(X, qfact.U, LV[1:(2k), 1:k])
     return X
 end
+
+@doc raw"""
+    Y = riemannian_Hessian(M::MetricManifold{ℝ, Stiefel{n,k}, CanonicalMetric}, p, G, H, X)
+    riemannian_Hessian!(M::MetricManifold{ℝ, Stiefel{n,k}, CanonicalMetric}, Y, p, G, H, X)
+
+Compute the Riemannian Hessian ``\operatorname{Hess} f(p)[X]`` given the
+Euclidean gradient ``∇ f(\tilde p)`` in `G` and the Euclidean Hessian ``∇^2 f(\tilde p)[\tilde X]`` in `H`,
+where ``\tilde p, \tilde X`` are the representations of ``p,X`` in the embedding,.
+
+Here, we adopt Eq. (5.6) [Nguyen:2023](@cite), for the [`CanonicalMetric`](@ref)
+``α_0=1, α_1=\frac{1}{2}`` in their formula. The formula reads
+
+```math
+    \operatorname{Hess}f(p)[X]
+    =
+    \operatorname{proj}_{T_p\mathcal M}\Bigl(
+        ∇^2f(p)[X] - \frac{1}{2} X \bigl( (∇f(p))^{\mathrm{H}}p + p^{\mathrm{H}}∇f(p)\bigr)
+        - \frac{1}{2} \bigl( P ∇f(p) p^{\mathrm{H}} + p ∇f(p))^{\mathrm{H}} P)X
+    \Bigr),
+```
+where ``P = I-pp^{\mathrm{H}}``.
+
+Compared to Eq. (5.6) we have ``α_0 = 1`` and ``α_1 = \frac{1}{2}``.
+"""
+riemannian_Hessian(
+    M::MetricManifold{𝔽,Stiefel{n,k,𝔽},CanonicalMetric},
+    p,
+    G,
+    H,
+    X,
+) where {n,k,𝔽}
+
+function riemannian_Hessian!(
+    M::MetricManifold{𝔽,Stiefel{n,k,𝔽},CanonicalMetric},
+    Y,
+    p,
+    G,
+    H,
+    X,
+) where {n,k,𝔽}
+    project!(
+        M,
+        Y,
+        p,
+        H - 1 / 2 .* X * (G' * p + p' * G) -
+        1 / 2 * ((I - p * p') * G * p' + p * G' * (I - p * p')) * X,
```

Comment:
`G' * p` is transposed `p' * G`. Similarly `((I - p * p') * G * p'`  seems to be transposition of `p * G' * (I - p * p'))`. By exploiting it we get fewer matrix multiplications.

Reply:
But (again) We need to compute them then before and allocate?